DataStructures/BinarySearchTree/BinarySearchTree.cs

```diff
@@ -0,0 +1,421 @@
+            return true;
+        }
+
+        /// <summary>
+        /// Returns a node with the smallest key.
+        /// </summary>
+        /// <returns>The node if possible, a default value otherwise.</returns>
+        public BinarySearchTreeNode<TKey>? GetMin()
+        {
+            if (root == null)
+            {
+                return default;
+            }
+
+            return GetMin(root);
+        }
+
+        /// <summary>
+        /// Returns a node with the largest key.
+        /// </summary>
+        /// <returns>The node if possible, a default value otherwise.</returns>
+        public BinarySearchTreeNode<TKey>? GetMax()
+        {
+            if (root == null)
+            {
+                return default;
+            }
+
+            return GetMax(root);
+        }
+
+        /// <summary>
+        /// Returns all the keys in the BST, sorted In-Order.
+        /// </summary>
+        /// <returns>A list of keys in the BST.</returns>
+        public ICollection<TKey> GetKeysInOrder() => GetKeysInOrder(root);
+
+        /// <summary>
+        /// Returns all the keys in the BST, sorted Pre-Order.
+        /// </summary>
+        /// <returns>A list of keys in the BST.</returns>
+        public ICollection<TKey> GetKeysPreOrder() => GetKeysPreOrder(root);
+
+        /// <summary>
+        /// Returns all the keys in the BST, sorted Post-Order.
+        /// </summary>
+        /// <returns>A list of keys in the BST.</returns>
+        public ICollection<TKey> GetKeysPostOrder() => GetKeysPostOrder(root);
+
+        public override string ToString() => $"BinarySearchTree<{typeof(TKey)}>: Count={Count}";
+
+        /// <summary>
+        /// Recursive method to add a key to the BST.
+        /// </summary>
+        /// <param name="node">Node to search from.</param>
+        /// <param name="key">Key to add.</param>
+        /// <exception cref="ArgumentException">
+        /// Thrown if key is already in the BST.
+        /// </exception>
+        private void Add(BinarySearchTreeNode<TKey> node, TKey key)
+        {
+            int compareResult = node.Key.CompareTo(key);
+            if (compareResult > 0)
+            {
+                if (node.Left != null)
+                {
+                    Add(node.Left, key);
+                }
+                else
+                {
+                    var newNode = new BinarySearchTreeNode<TKey>(key);
+                    newNode.Parent = node;
+                    node.Left = newNode;
+                }
+            }
+            else if (compareResult < 0)
+            {
+                if (node.Right != null)
+                {
+                    Add(node.Right, key);
+                }
+                else
+                {
+                    var newNode = new BinarySearchTreeNode<TKey>(key);
+                    newNode.Parent = node;
+                    node.Right = newNode;
+                }
+            }
+            else if (compareResult == 0)
```

Comment:
Codacy found an issue: [Add the missing 'else' clause.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6277689)